samza-test/src/test/java/org/apache/samza/test/processor/TestZkLocalApplicationRunner.java

```diff
@@ -0,0 +1,427 @@
+  }
+
+  // Checks enforcing property that all processors should have unique Id.
+  // Depends upon SAMZA-1302
+  // @Test(expected = Exception.class)
+  public void shouldFailWhenNewProcessorJoinsWithSameIdAsExistingProcessor() throws InterruptedException {
+    // Set up kafka topics.
+    publishKafkaEvents(inputKafkaTopic, NUM_KAFKA_EVENTS, PROCESSOR_IDS[0]);
+
+    // Create StreamApplications.
+    CountDownLatch kafkaEventsConsumedLatch = new CountDownLatch(NUM_KAFKA_EVENTS);
+    CountDownLatch processedMessagesLatch1 = new CountDownLatch(1);
+    CountDownLatch processedMessagesLatch2 = new CountDownLatch(1);
+
+    StreamApplication streamApp1 = new TestStreamApplication(inputKafkaTopic, outputKafkaTopic, processedMessagesLatch1, null, kafkaEventsConsumedLatch);
+    StreamApplication streamApp2 = new TestStreamApplication(inputKafkaTopic, outputKafkaTopic, processedMessagesLatch2, null, kafkaEventsConsumedLatch);
+
+    // Run stream applications.
+    applicationRunner1.run(streamApp1);
+    applicationRunner2.run(streamApp2);
+
+    // Wait for message processing to start in both the processors.
+    processedMessagesLatch1.await();
+    processedMessagesLatch2.await();
+
+    LocalApplicationRunner applicationRunner3 = new LocalApplicationRunner(new MapConfig(applicationConfig2));
+
+    // Create a stream app with same processor id as SP2 and run it. It should fail.
+    publishKafkaEvents(inputKafkaTopic, NUM_KAFKA_EVENTS, PROCESSOR_IDS[2]);
+    kafkaEventsConsumedLatch = new CountDownLatch(NUM_KAFKA_EVENTS);
+    StreamApplication streamApp3 = new TestStreamApplication(inputKafkaTopic, outputKafkaTopic, null, null, kafkaEventsConsumedLatch);
+    applicationRunner3.run(streamApp3);
+
+    // The following line should throw up by handling duplicate processorId registration.
+    kafkaEventsConsumedLatch.await();
+  }
+
+  // Depends upon SAMZA-1302
+  // @Test(expected = Exception.class)
+  public void shouldKillStreamAppWhenZooKeeperDiesBeforeLeaderReElection() throws InterruptedException {
+    // Create StreamApplications.
+    StreamApplication streamApp1 = new TestStreamApplication(inputKafkaTopic, outputKafkaTopic, null, null, null);
+    StreamApplication streamApp2 = new TestStreamApplication(inputKafkaTopic, outputKafkaTopic, null, null, null);
+
+    // Run stream applications.
+    applicationRunner1.run(streamApp1);
+    applicationRunner2.run(streamApp2);
+
+    applicationRunner1.kill(streamApp1);
+    applicationRunner1.waitForFinish();
+    assertEquals(ApplicationStatus.SuccessfulFinish, applicationRunner1.status(streamApp2));
+
+    // Kill zookeeper server and expect job model regeneration and ZK fencing will fail with exception.
+    zookeeper().shutdown();
+
+    applicationRunner1 = new LocalApplicationRunner(applicationConfig1);
+    applicationRunner1.run(streamApp1);
+    // This line should throw exception since Zookeeper is unavailable.
+    applicationRunner1.waitForFinish();
```

Comment:
When zookeeper is shutdown, all the processors should also terminate. Please add those assertions when you get to it after bugfix. Thanks :)